Renovation of contribution guidelines & github templates

Ward9250 opened the pull request:
This revamps the contribution guidelines, as a document which can be included in every repository BioJulia manages, and is [recognised by github](https://help.github.com/articles/setting-guidelines-for-repository-contributors/).

This also introduces some templates for [issues](https://help.github.com/articles/creating-an-issue-template-for-your-repository/) and [pull-requests](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/) to help make keeping to the guidelines easy.

I doubt using git submodules will work and get github to notice the templates and such in repos. So I will have to make a tool that deploys these files to BioJulia repositories automatically.

In the meantime, feedback on these guidelines is welcome.

Ward9250 commented:
@ScotPJones, I've implemented your suggestions from the BioJulia/BioCore.jl PR. #

ScottPJones commented on CONTRIBUTING.md:
```diff
@@ -574,13 +570,10 @@ include("feature2.jl")
+- Files that declare modules should have the same name name of the module.
   E.g the module `SomeModule` is declared under the file `SomeModule.jl`.
 
-* Separate logical blocks of code with one blank line,
```
I didn't mean that this should totally disappear, in particular, the separation of logical blocks of code with one blank line is a good guideline.

ScottPJones commented on CONTRIBUTING.md:
````diff
@@ -601,7 +594,7 @@ name(node) # get node name
 ```
 
-* When using conditional branching, if code is statement-like, an
+- When using conditional branching, if code is statement-like, an
````
A question about this, do you mean that one should not use the common Julia idiom:
```julia
   expression && statement
```
or
```julia
    expression &&
        statement
```
and the same but with `||`.
This is very common for simple checks, such as `isvalid(foo) || throw(ArgumentError("$foo is not valid"))`

jgreener64 commented:
I think these changes are good, thank for putting in the work.

The only reservation I have is that the issue and PR templates are quite extensive. Whilst this makes it easier for maintainers to deal with individual issues/PRs, I wonder if it might act to turn people away from reporting bugs or quick enhancements.

I know some projects use templates to good effect and I agree they make things easier - does anyone know of a project that uses such extensive ones as here?

Ward9250 commented:
@ScottPJones I've changed `findletter`, added
> For functions that are not a single expression, it is preferred to use an explicit `return`.
> Be aware that functions in julia implicitly return the the result of the last expression in the function, > so plain `return` should be used to indicate that the function returns `nothing`.

Ward9250 commented:
@jgreener64 I wonder about that as well. Perhaps we should put inside them something like *This template is rather extensive. Fill out all that you can, if are a new contributor or you're unsure about anything, leave it unchanged and a reviewer will help you :smile: *

Ward9250 commented:
I should also say, these files are a standard BioJulia wide suggestion, and are the best files I could come up with in a fortnight. That doesent make them perfect, and it doesent make them fully applicable to every repo, or the last word on contributing.

If you've been watching BioCore.jl you'll note an odd pull request - that was generated by the rollout tool I prototyped yesterday - rather than simply pushing these files to all the repos rather dictatorially, I've designed it to create a branch, update the file(s) if they already exist, and then make a PR. Because we have a decentralised maintainer model, and these files are something the maintainers of individual repos should have a say in. Rolling them out that way allows everyone to merge and resolve what they want, and customise these files on their own repositories - I hope! 😆

ScottPJones commented:
Thanks for being so responsive to the review suggestions!

ScottPJones commented:
Note: one thing that is totally new to me is the idea of using emoji's on commit messages.  Where did that come from?  I still have to get used to the usage, and teaching my brain to recognize the meanings :grinning:

Ward9250 commented:
No problem at all, I see it as a primary responsibility to make sure you all happy working on and using BioJulia. Yeah the emoji's blew my mind! But also, even if you're using something that's not GitHub or something that can render them, their syntax can make doing stuff like grepping logs and things pretty nice!
